source/simulator/assembly.cc

```diff
@@ -1004,6 +1002,276 @@ namespace aspect
+                                     // we assemble the matrix)
+                                     (pressure_scaling *
+                                      compressibility * density *
+                                      (scratch.velocity_values[q] * gravity) *
+                                      scratch.phi_p[i])
+                                   )
+                                   * scratch.finite_element_values.JxW(q);
+          }
+      }
+
+
+
+      template <int dim>
+      void
+      local_assemble_stokes_system_incompressible (const SimulatorAccess<dim>                      &simulator_access,
+                                                   const double                                     pressure_scaling,
+                                                   const bool                                       rebuild_stokes_matrix,
+                                                   internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                   internal::Assembly::CopyData::StokesSystem<dim> &data)
+      {
+        const Introspection<dim> &introspection = simulator_access.introspection();
+        const unsigned int dofs_per_cell = scratch.finite_element_values.get_fe().dofs_per_cell;
+        const unsigned int n_q_points    = scratch.finite_element_values.n_quadrature_points;
+
+        for (unsigned int q=0; q<n_q_points; ++q)
+          {
+            for (unsigned int k=0; k<dofs_per_cell; ++k)
+              {
+                scratch.phi_u[k] = scratch.finite_element_values[introspection.extractors.velocities].value (k,q);
+                scratch.phi_p[k] = scratch.finite_element_values[introspection.extractors.pressure].value (k, q);
+                if (rebuild_stokes_matrix)
+                  {
+                    scratch.grads_phi_u[k] = scratch.finite_element_values[introspection.extractors.velocities].symmetric_gradient(k,q);
+                    scratch.div_phi_u[k]   = scratch.finite_element_values[introspection.extractors.velocities].divergence (k, q);
+                  }
+              }
+
+            // Viscosity scalar
+            const double eta = (rebuild_stokes_matrix
+                                ?
+                                scratch.material_model_outputs.viscosities[q]
+                                :
+                                std::numeric_limits<double>::quiet_NaN());
+
+            const SymmetricTensor<4,dim> &stress_strain_director =
+              scratch.material_model_outputs.stress_strain_directors[q];
+            const bool use_tensor = (stress_strain_director !=  dealii::identity_tensor<dim> ());
+
+            const Tensor<1,dim>
+            gravity = simulator_access.get_gravity_model().gravity_vector (scratch.finite_element_values.quadrature_point(q));
+
+            const double density = scratch.material_model_outputs.densities[q];
```

Comment:
I just copied the existing code. No functional changes are intended in this pull request.